master/buildbot/test/util/integration.py

```diff
@@ -21,16 +21,65 @@
 
 from twisted.internet import defer
 from twisted.internet import reactor
+from twisted.internet import task
 from twisted.trial import unittest
 
 from buildbot.master import BuildMaster
 from buildbot.status.results import SUCCESS
 from buildbot.status.results import statusToString
 from buildbot.test.util import dirs
+from buildbot.schedulers.basic import BaseBasicScheduler
 from buildslave.bot import BuildSlave
 from buildslave.bot import LocalBuildSlave
 
 
+class WaitForStartConsumingSchedulers(object):
+    """
+    startService does not return a deferred. So, when
+    we call yield master.reconfigServiceWithBuildbotConfig
+    we cannot know whether services, like schedulers, have finished their
+    setup.
+    This creates an issue regarding the SingleBranch, AnyBranch schedulers.
+    These schedulers are able to trigger builds only after they have
+    subscribed to the 'new' change event. If we try to start a build with a new change
+    before they subscribed to the event (what can happen as startService does not return a deferred),
+    the build will never start, and the test will timeout.
+    This class implements a poller to ensure that these schedulers have subscribed to the event,
+    before any test tries to create a build with a new change.
+    """
+    POLL_INTERVAL_SEC = 1
```

Comment:
ClusteredService is for multimaster. I dont see a reason why is startService could not be made async. 
I have no strong opinion about rushing into fixing this racecondition, as this is a known flaky test.

Our current way about flaky tests is to mark them flaky with a decorator, which effectively put them out of the CI. I am not a big fan of this latter option as we ends up just forgotting about those test, and eventually make them fail 100%.

so go ahead and merge, I would prefer merging a workaround rather than disabling the test.